qiskit_experiments/library/characterization/cr_hamiltonian.py

```diff
@@ -0,0 +1,472 @@
+                        risefall_sigma_ratio=opt.risefall,
+                    ),
+                    pulse.drive_channel(self.physical_qubits[1]),
+                )
+            else:
+                pulse.delay(duration, pulse.drive_channel(self.physical_qubits[1]))
+
+            # place holder for empty drive channels. this is necessary due to known pulse gate bug.
+            pulse.delay(duration, pulse.drive_channel(self.physical_qubits[0]))
+
+        with pulse.build(backend, default_alignment="left", name="cr_m") as cross_resonance_m:
+            with pulse.phase_offset(
+                np.pi,
+                pulse.control_channels(*self.physical_qubits)[0],
+                pulse.drive_channel(self.physical_qubits[1]),
+            ):
+                # add cross resonance tone
+                pulse.play(
+                    pulse.GaussianSquare(
+                        duration,
+                        amp=opt.amp,
+                        sigma=sigma,
+                        risefall_sigma_ratio=opt.risefall,
+                    ),
+                    pulse.control_channels(*self.physical_qubits)[0],
+                )
+                # add cancellation tone
+                if not np.isclose(opt.amp_t, 0.0):
+                    pulse.play(
+                        pulse.GaussianSquare(
+                            duration,
+                            amp=opt.amp_t,
+                            sigma=sigma,
+                            risefall_sigma_ratio=opt.risefall,
+                        ),
+                        pulse.drive_channel(self.physical_qubits[1]),
+                    )
+                else:
+                    pulse.delay(duration, pulse.drive_channel(self.physical_qubits[1]))
+
+            # place holder for empty drive channels. this is necessary due to known pulse gate bug.
+            pulse.delay(duration, pulse.drive_channel(self.physical_qubits[0]))
+
+        cr_circuit.add_calibration(
+            gate=cr_gate_p,
+            qubits=self.physical_qubits,
+            schedule=cross_resonance_p,
+            params=[duration],
+        )
+        cr_circuit.add_calibration(
+            gate=cr_gate_m,
+            qubits=self.physical_qubits,
+            schedule=cross_resonance_m,
+            params=[duration],
+        )
+
+        return cr_circuit
+
+    def _net_duration(
+        self,
+        duration: float,
+        sigma: float,
+    ) -> float:
+        """Calculate net cross resonance pulse duration considering the pulse edges.
+
+        Args:
+            duration: Parameter object representing a duration of cross resonance pulse.
+            sigma: Sigma of Gaussian edges.
+
+        Returns:
+            Net duration in units of sec.
+        """
+        opt = self.experiment_options
+
+        flat_top_width = duration - 2 * opt.risefall * sigma
+        net_edge_width = np.sqrt(2 * np.pi * sigma ** 2)
+
+        return 2 * (flat_top_width + net_edge_width)
```

Comment:
You could probably save yourself this piece of code by defining a class variable `__n_pulses__` which would be 1 for CrossResonanceHamiltonian`  and 2 for the echoed variant.

Reply:
I prefer explicitly writing this method for each child class because we can easily understand how we compute xval for defined decomposition, without understanding complicated logic. Of course some child class may implement fancy pulse sequence, which doesn't repeat the same pulse multiple times.